source/extensions/priv/server/ntds_jet.c

```diff
@@ -0,0 +1,432 @@
+	if (shutdownStatus != JET_errSuccess){
+		return shutdownStatus;
+	}
+	shutdownStatus = JetDetachDatabase(ntdsState->jetSession, ntdsState->ntdsPath);
+	if (shutdownStatus != JET_errSuccess){
+		return shutdownStatus;
+	}
+	shutdownStatus = JetEndSession(ntdsState->jetSession, (JET_GRBIT)NULL);
+	if (shutdownStatus != JET_errSuccess){
+		return shutdownStatus;
+	}
+	shutdownStatus = JetTerm(ntdsState->jetEngine);
+	free(ntdsState);
+	return shutdownStatus;
+}
+
+/*!
+* @brief Starts up the Jet Instance and initialises it.
+* @param ntdsState Pointer to a jetsState struct which contains all the state data for the Jet Instance.
+* @returns Indication of sucess or failure.
+*/
+JET_ERR engine_startup(struct jetState *ntdsState){
+	JET_ERR jetError;
+	// Set the Page Size to the highest possibile limit
+	jetError = JetSetSystemParameter(&ntdsState->jetEngine, JET_sesidNil, JET_paramDatabasePageSize, 8192, NULL);
+	if (jetError != JET_errSuccess){
+		return jetError;
+	}
+	char instanceName[80] = "NTDS ";
+	get_instance_name(instanceName);
+	// Create our Jet Instance
+	jetError = JetCreateInstance(&ntdsState->jetEngine, instanceName);
+	if (jetError != JET_errSuccess){
+		return jetError;
+	}
+	// Disable crash recovery and transaction logs
+	jetError = JetSetSystemParameter(&ntdsState->jetEngine, JET_sesidNil, JET_paramRecovery, (JET_API_PTR)NULL, "Off");
+	if (jetError != JET_errSuccess){
+		return jetError;
+	}
+	// Initialise the Jet instance
+	jetError = JetInit(&ntdsState->jetEngine);
+	if (jetError != JET_errSuccess){
+		return jetError;
+	}
+	return JET_errSuccess;
+}
+
+void get_instance_name(char *name){
+	SYSTEMTIME currentTime;
+	GetSystemTime(&currentTime);
+	char dateString[30];
+	char timeString[30];
+	GetDateFormat(LOCALE_SYSTEM_DEFAULT, DATE_LONGDATE, &currentTime, NULL, dateString, 255);
+	strncat(name, dateString, sizeof(dateString));
+	GetTimeFormat(LOCALE_SYSTEM_DEFAULT, 0, &currentTime, NULL, timeString, 255);
+	strncat(name, timeString, sizeof(timeString));
+}
+
+
+/*!
+* @brief Moves the database cursor to the first record in the 'datatable' table
+* @param ntdsState Pointer to a jetsState struct which contains all the state data for the Jet Instance.
+* @returns Indication of sucess or failure.
+*/
+JET_ERR find_first(struct jetState *ntdsState){
+	JET_ERR cursorStatus;
+	cursorStatus = JetMove(ntdsState->jetSession, ntdsState->jetTable, JET_MoveFirst, (JET_GRBIT)NULL);
+	return cursorStatus;
+}
+
+/*!
+* @brief Collect the Column Definitions for all relevant columns in 'datatable'
+* @param ntdsState Pointer to a jetsState struct which contains all the state data for the Jet Instance.
+* @param accountColumns Pointer to an ntdsState struct which will hold all of our column definitions.
+* @returns Indication of sucess or failure.
+*/
+JET_ERR get_column_info(struct jetState *ntdsState, struct ntdsColumns *accountColumns){
+	JET_ERR columnError;
+	struct {

```

Comment:
I suggested it. Too clever? I think its easier to read than the previous version, which traversed two parallel arrays. Now the names and columns are self-documenting since they are on the same line.